src/components/CustomCheckout.tsx

```diff
@@ -0,0 +1,271 @@
+  if (!customCheckoutSdk) {
+    return null;
+  }
+
+  const {on: _on, session: _session, ...actions} = customCheckoutSdk;
+  if (!sessionState) {
+    return {...actions, ...customCheckoutSdk.session()};
+  }
+
+  return {...actions, ...sessionState};
+};
+
+interface CustomCheckoutProviderProps {
+  /**
+   * A [Stripe object](https://stripe.com/docs/js/initializing) or a `Promise` resolving to a `Stripe` object.
+   * The easiest way to initialize a `Stripe` object is with the the [Stripe.js wrapper module](https://github.com/stripe/stripe-js/blob/master/README.md#readme).
+   * Once this prop has been set, it can not be changed.
+   *
+   * You can also pass in `null` or a `Promise` resolving to `null` if you are performing an initial server-side render or when generating a static site.
+   */
+  stripe: PromiseLike<stripeJs.Stripe | null> | stripeJs.Stripe | null;
+  options: stripeJs.StripeCustomCheckoutOptions;
+}
+
+interface PrivateCustomCheckoutProviderProps {
+  stripe: unknown;
+  options: stripeJs.StripeCustomCheckoutOptions;
+  children?: ReactNode;
+}
+const INVALID_STRIPE_ERROR =
+  'Invalid prop `stripe` supplied to `CustomCheckoutProvider`. We recommend using the `loadStripe` utility from `@stripe/stripe-js`. See https://stripe.com/docs/stripe-js/react#elements-props-stripe for details.';
+
+export const CustomCheckoutProvider: FunctionComponent<PropsWithChildren<
+  CustomCheckoutProviderProps
+>> = (({
+  stripe: rawStripeProp,
+  options,
+  children,
+}: PrivateCustomCheckoutProviderProps) => {
+  const parsed = React.useMemo(
+    () => parseStripeProp(rawStripeProp, INVALID_STRIPE_ERROR),
+    [rawStripeProp]
+  );
+
+  // State used to trigger a re-render when sdk.session is updated
+  const [
+    session,
+    setSession,
+  ] = React.useState<stripeJs.StripeCustomCheckoutSession | null>(null);
+
+  // State used to avoid calling initCustomCheckout multiple times when options changes
+  const [
+    initCustomCheckoutCalled,
+    setInitCustomCheckoutCalled,
+  ] = React.useState<boolean>(false);
+
+  const [ctx, setContext] = React.useState<CustomCheckoutSdkContextValue>(
+    () => ({
+      stripe: parsed.tag === 'sync' ? parsed.stripe : null,
+      customCheckoutSdk: null,
+    })
+  );
+
+  const safeSetContext = (
+    stripe: stripeJs.Stripe,
+    customCheckoutSdk: stripeJs.StripeCustomCheckout
+  ) => {
+    setContext((ctx) => {
+      if (ctx.stripe && ctx.customCheckoutSdk) {
+        return ctx;
+      }
+
+      return {stripe, customCheckoutSdk};
+    });
+  };
+
+  React.useEffect(() => {
+    let isMounted = true;
+
+    if (parsed.tag === 'async' && !ctx.stripe) {
+      parsed.stripePromise.then((stripe) => {
+        if (stripe && isMounted && !initCustomCheckoutCalled) {
+          // Only update context if the component is still mounted
+          // and stripe is not null. We allow stripe to be null to make
+          // handling SSR easier.
+          setInitCustomCheckoutCalled(true);
```

Comment:
Thoughts on using a ref for this instead of state? Do we need to recompute effects/render when this value changes, or just use it to prevent multiple invocations of initCustomCheckout?

Reply:
It was just to prevent multiple invocations of `initCustomCheckout`. 
 Let me give a try of using ref.

Reply:
updated https://github.com/stripe/react-stripe-js/pull/439/commits/195b66f0a0e403d04c053f333635fb86620890b4